---
title: "Manuscript Figures"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval.after = 'fig.cap')
```

```{r importSynMapData, include=FALSE}
source("./R/do.R")
source("./R/pathway_analysis.R")

library(ggplot2)
library(tidyr)
library(dplyr)
library(VennDiagram)
library(gridExtra)
```

Some direct numbers for reporting in various parts of the manuscript.

***  

Overlap in reactions and pathways annotated to retained pairs.

```{r numbers1, echo=FALSE, results='asis'}
knitr::kable(getPathwayReactionOverlap(homeologs.pairs) %>% select(Description, Value))
```

How often does the maize1 gene express higher than maize2?  Only count when both genes have expression.

```{r numbers2, echo=FALSE, results='asis'}
data <- 
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>% #only matching
  select(Maize1, Maize2) %>%
  inner_join(maize.walley.v4mapped.expression, by=c("Maize1"="geneID")) %>%
  rename(FPKM_avg1=FPKM_avg) %>%
  inner_join(maize.walley.v4mapped.expression, by=c("Maize2"="geneID", "sample"="sample")) %>%
  rename(FPKM_avg2=FPKM_avg) %>%
  select(Maize1, Maize2, sample, FPKM_avg1, FPKM_avg2) %>%
  count(sample, sign=sign(FPKM_avg1-FPKM_avg2)) %>%
  spread(key = sign, value = n) %>%
  mutate(percent_favor_m1=`1`/(`-1`+`1`)) %>%
  rename(m2_dom=`-1`, m1_dom=`1`) %>%
  select(sample, m1_dom, m2_dom, percent_favor_m1) %>%
  arrange(percent_favor_m1)
knitr::kable(data)
```

 * There are **`r nrow(homeologs.pairs %>% subset(Maize1 != "" & Maize2 != ""))`** retained duplicate pairs.
 * 
 <!-- * Summary of the KS values -->
 <!--     + **`r 1+2`** -->

# 1 Chromosome location of Maize1 and Maize2

***

```{r chromosome_plot_exp, echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=10, fig.height=5}
subgenome.positions <-
  subgenome %>%
  left_join(B73.v4.gene.positions, by=c("gene2"="geneID")) %>%
  select(Chromosome, Pos_start, Pos_end, subgenome) %>%
  rename(Subgenome=subgenome) %>%
  subset(Subgenome %in% c("sub1", "sub2"))

dominant_homeolog_exp <- expressedPairs
dominant_homeolog_exp$Subgenome <- "no dominance"
dominant_homeolog_exp$Subgenome[dominant_homeolog_exp$FPKM_mean1 / dominant_homeolog_exp$FPKM_mean2 > 2] <- "Dominant Expression"
dominant_homeolog_exp$Subgenome[dominant_homeolog_exp$FPKM_mean2 / dominant_homeolog_exp$FPKM_mean1 > 2] <- "Dominant Expression"
dominant_homeolog_exp$Maize1[dominant_homeolog_exp$FPKM_mean2 / dominant_homeolog_exp$FPKM_mean1 > 2] <- dominant_homeolog_exp$Maize2[dominant_homeolog_exp$FPKM_mean2 / dominant_homeolog_exp$FPKM_mean1 > 2]
df0 <- dominant_homeolog_exp[dominant_homeolog_exp$Subgenome %in% c("no dominance"),]
df0$Maize1 <- df0$Maize2
dominant_homeolog_exp <- bind_rows(dominant_homeolog_exp, df0)
dominant_homeolog_exp <-
  dominant_homeolog_exp %>%
  select(Maize1, Subgenome) %>%
  left_join(B73.v4.gene.positions, by=c("Maize1"="geneID")) %>%
  select(Chromosome, Pos_start, Pos_end, Subgenome)

df1 <- B73.v4.centromere.positions
df1$x <- -.5
df1$xend <- .05
df2 <- subgenome.positions
df2$x <- -.45
df2$xend <- -.05
df3 <- dominant_homeolog_exp
df3$x <- .05
df3$xend <- .25
segment_data <- bind_rows(df2,df3)
segment_data <- segment_data[!is.na(segment_data$Pos_start),]
segment_data <-
  segment_data %>%
  ungroup()
segment_data$Subgenome[segment_data$Subgenome %in% c("sub1")] <- "Maize1"
segment_data$Subgenome[segment_data$Subgenome %in% c("sub2")] <- "Maize2"
# segment_data$Subgenome[segment_data$Subgenome %in% c("dom")] <- "Dominant Homeolog"
segment_data <- segment_data %>% subset(Subgenome %in% c("Maize1", "Maize2", "Dominant Expression"))
centromere_data <- df1

## Make the plot
chromosome_plot_exp <-
  ggplot(data=B73.v4.chr.size) +
  geom_bar(aes(B73.v4.chr.size$chromosome, B73.v4.chr.size$size), stat="identity", fill="grey70") +
  # geom_segment(data=segment_data, aes(x=Chromosome+x, xend=Chromosome+xend, y=Pos_start, yend=Pos_end, colour=Subgenome), size=.001) +
  geom_rect(data=segment_data, inherit.aes = F, aes(xmin=Chromosome+x, xmax=Chromosome+xend, ymin=Pos_start, ymax=Pos_end, color=Subgenome, fill=Subgenome)) +
  geom_rect(data=centromere_data, inherit.aes = F, aes(xmin=Chromosome+x, xmax=Chromosome+xend, ymin=Pos_start, ymax=Pos_end)) +
  scale_fill_manual(values=c("green", "red", "blue", "black")) +
  scale_color_manual(values=c("green", "red", "blue", "black")) +
  scale_x_discrete(limits=c("Chr1","Chr2","Chr3","Chr4","Chr5","Chr6","Chr7","Chr8","Chr9","Chr10")) +
  scale_y_continuous(labels=format_si())
chromosome_plot_exp  +
  labs(
    # title="Subgenome Locations by Chromosome",
    x="Chromosome",
    y="Position on Chromosome"
  ) + coord_flip()
```

# 2 GO Term Comparison by Subgenome

***

```{r GOTerms_by_subgenome, echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=4, fig.height=4}
# #EXP only, MF terms only
# grid.newpage()
# GOTerms_by_subgenome_a <- draw.pairwise.venn(
#   area1 = goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct() %>% nrow(),
#   area2 = goAnnotations.sub2 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct() %>% nrow(),
#   cross.area = intersect(goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct(),  
#                          goAnnotations.sub2 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct()
#                          ) %>% nrow(), 
#   category = c("Maize1", "Maize2"), 
#   lty = rep("blank", 2), 
#   fill = c("light blue", "green"), 
#   alpha = rep(0.5, 2), 
#   cat.pos = c(0, 180), 
#   scaled = FALSE,
#   euler.d = FALSE, 
#   sep.dist = 0.3, 
#   rotation.degree = 0,
#   ind = FALSE
# )
# grid.arrange(gTree(children=GOTerms_by_subgenome_a), top="GO Term by Subgenome\n(EV-EXP, MF only)")
# 
# ## Only keep genes where there is a duplicate still on subgenome 1 and subgenome 2.  The converted data.frame will hold maize2 genes with the names replaced with maize1 homologs
# converted <-
#   homeologs.pairs %>%
#   subset(Maize1 != "" & Maize2 != "") %>%
#   right_join(., goAnnotations.sub2,by=c("Maize2"="gene2"))
# converted$Maize2[!is.na(converted$Maize1)] <- as.character(converted$Maize1[!is.na(converted$Maize1)])
# converted <- 
#   converted %>%
#   select(Maize2, goTerm, evCode, type) %>%
#   rename(gene2=Maize2)
# 
# #EXP only, MF only
# grid.newpage()
# GOTerms_by_subgenome_b <- draw.pairwise.venn(
#   area1 = goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% nrow(),
#   area2 = converted %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% nrow(),
#   cross.area = intersect(goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(gene2, goTerm) %>% distinct(), 
#                          converted %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(gene2, goTerm) %>% distinct()
#                          ) %>% nrow(), 
#   category = c("Maize1", "Maize2"), 
#   lty = rep("blank", 2), 
#   fill = c("light blue", "green"), 
#   alpha = rep(0.5, 2), 
#   cat.pos = c(0, 180), 
#   scaled = FALSE,
#   euler.d = FALSE, 
#   sep.dist = 0.3, 
#   rotation.degree = 0,
#   ind = FALSE
# )
# grid.arrange(gTree(children=GOTerms_by_subgenome_b), top="GO Term Assignments by Subgenome\n(EV-EXP, MF only)")

#EXP only, MF terms only
grid.newpage()
GOTerms_by_subgenome_a <- draw.pairwise.venn(
  area1 = goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct() %>% nrow(),
  area2 = goAnnotations.sub2 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct() %>% nrow(),
  cross.area = intersect(goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct(),  
                         goAnnotations.sub2 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(goTerm) %>% distinct()
                         ) %>% nrow(), 
  category = c("Maize1", "Maize2"), 
  lty = rep("blank", 2), 
  fill = c("light blue", "green"), 
  alpha = rep(0.5, 2), 
  cat.pos = c(0, 180), 
  scaled = FALSE,
  euler.d = FALSE, 
  sep.dist = 0.3, 
  rotation.degree = 0,
  ind = FALSE
)
grid.arrange(gTree(children=GOTerms_by_subgenome_a), top="GO Term by Subgenome\n(EV-EXP, MF only)")

## Only keep genes where there is a duplicate still on subgenome 1 and subgenome 2.  The converted data.frame will hold maize2 genes with the names replaced with maize1 homologs
converted <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  right_join(., goAnnotations.sub2,by=c("Maize2"="gene2"))
converted$Maize2[!is.na(converted$Maize1)] <- as.character(converted$Maize1[!is.na(converted$Maize1)])
converted <- 
  converted %>%
  select(Maize2, goTerm, evCode, type) %>%
  rename(gene2=Maize2)

#EXP only, MF only
grid.newpage()
GOTerms_by_subgenome_b <- draw.pairwise.venn(
  area1 = goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% nrow(),
  area2 = converted %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% nrow(),
  cross.area = intersect(goAnnotations.sub1 %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(gene2, goTerm) %>% distinct(), 
                         converted %>% filter(evCode=="EXP") %>% filter(type=="MF") %>% select(gene2, goTerm) %>% distinct()
                         ) %>% nrow(), 
  category = c("Maize1", "Maize2"), 
  lty = rep("blank", 2), 
  fill = c("light blue", "green"), 
  alpha = rep(0.5, 2), 
  cat.pos = c(0, 180), 
  scaled = FALSE,
  euler.d = FALSE, 
  sep.dist = 0.3, 
  rotation.degree = 0,
  ind = FALSE
)
grid.arrange(gTree(children=GOTerms_by_subgenome_b), top="GO Term Assignments by Subgenome\n(EV-EXP, MF only)")
```


# 3 Isoforms

***

Here is a scatter plot of isoform counts across gene pairs.

```{r subgenome_isoforms_scatter, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align='center'}
# Isoforms
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(geneTranscript.counts, by=c("Maize1"="gene")) %>%
  inner_join(geneTranscript.counts, by=c("Maize2"="gene"))
names(data)[3] <- "Value_maize1"
names(data)[4] <- "Value_maize2"
# data <- subset(data, data$Value_maize1 > 1 & data$Value_maize2 > 1) ## Ignore single-exon genes
data$foldChange <- log2(data$Value_maize1) - log2(data$Value_maize2)

n <-
  data %>%
  # subset(Value_maize1 > 100 | Value_maize2 > 100) %>%
  subset(Value_maize1 > 25 | Value_maize2 > 25) %>%
  nrow()

subgenome_isoforms_scatter <-
  ggplot(subset(data, Value_maize1 <= 25 & Value_maize2 <= 25), aes(x=Value_maize1, y=Value_maize2)) +
  geom_point(position = "jitter", alpha=.1, aes(color = abs(foldChange))) +
  scale_color_gradient(low = "blue", high = "red") +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  # xlim(0,100) +
  # ylim(0,100) +
  xlim(0,25) +
  ylim(0,25) +
  labs(
    # title = paste0("Scatterplot of isoform counts for gene pair(s).\nExcludes ",n," outlier, alpha=.1, foldchange in log2, blue=lm fit."),
    x = "Isoform Counts for Maize1 Gene",
    y = "Isoform Counts for Maize2 Gene"
  ) + 
  # annotate("text",x=85,y=20, label=paste("Pearson Cor = ", round(cor(data$Value_maize1, data$Value_maize2, method = c("pearson")),2)))
  annotate("text",x=20,y=3, label=paste("Pearson Cor = ", round(cor(data$Value_maize1, data$Value_maize2, method = c("pearson")),2)))
plot(subgenome_isoforms_scatter)

```

# 4 Expression

***

```{r subgenome_expression_scatter_rep, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=2.5, fig.align='center'}
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(FPKM_avg.x) & !is.na(FPKM_avg.y))
names(data)[4] <- "Value_maize1"
names(data)[5] <- "Value_maize2"
# data <- subset(data, sample %in% c("Silk", "Embryo 20 DAP", "Leaf Zone 2 (Stomatal)"))

data <- 
  data %>%
  subset(sample %in% c("Primary Root 5 Days", "Mature Leaf 8", "6-7 internode", "B73 Mature Pollen"))

subgenome_expression_scatter_rep <-
  ggplot(data, aes(x=Value_maize1, y=Value_maize2)) +
  geom_point(alpha=.1, aes(color = sample)) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  scale_x_log10() + scale_y_log10() +
  labs(
    # title = paste0("Scatterplot of expression for gene pair(s).\nAlpha=.1, reference line"),
    x = "Expression for Maize1 Gene (log scale)",
    y = "Expression for Maize2 Gene (log scale)"
  ) +
  theme(legend.position = "none") + facet_wrap(~sample, nrow=1)
plot(subgenome_expression_scatter_rep)

```

```{r subgenome_expression_scatter, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=8, fig.align='center'}
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(FPKM_avg.x) & !is.na(FPKM_avg.y))
names(data)[4] <- "Value_maize1"
names(data)[5] <- "Value_maize2"
# data <- subset(data, sample %in% c("Silk", "Embryo 20 DAP", "Leaf Zone 2 (Stomatal)"))

subgenome_expression_scatter <-
  ggplot(data, aes(x=Value_maize1, y=Value_maize2)) +
  geom_point(alpha=.1, aes(color = sample)) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  scale_x_log10() + scale_y_log10() +
  labs(
    # title = paste0("Scatterplot of expression for gene pair(s).\nAlpha=.1, reference line"),
    x = "Expression for Maize1 Gene (log scale)",
    y = "Expression for Maize2 Gene (log scale)"
  ) +
  theme(legend.position = "none") + facet_wrap(~sample, nrow=6)
plot(subgenome_expression_scatter)
```

# 5 Protein Abundance

***

```{r subgenome_protein_scatter_rep, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=2.5, fig.align='center'}
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(dNSAF_avg.x) & !is.na(dNSAF_avg.y))
names(data)[4] <- "Value_maize1"
names(data)[5] <- "Value_maize2"
# data <- subset(data, sample %in% c("Silk", "Embryo 20 DAP", "Leaf Zone 2 (Stomatal)"))
data <- 
  data %>%
  subset(sample %in% c("Primary Root 5 Days", "Mature Leaf 8", "6-7 internode", "B73 Mature Pollen"))

subgenome_protein_scatter_rep <-
  ggplot(data, aes(x=Value_maize1, y=Value_maize2)) +
  geom_point(alpha=.1, aes(color = sample)) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  scale_x_log10() + scale_y_log10() +
  labs(
    # title = paste0("Scatterplot of protein abundance for gene pair(s).\nAlpha=.1, reference line"),
    x = "Abundance for Maize1 Gene (log scale)",
    y = "Abundance for Maize2 Gene (log scale)"
  ) +
  theme(legend.position = "none") + facet_wrap(~sample, nrow=1)
  # + annotate("text", size=2, x=10000,y=10, label=paste("pearson: ", round(cor(data$Value_maize1, data$Value_maize2, method = c("pearson")),2)))
plot(subgenome_protein_scatter_rep)
```

```{r subgenome_protein_scatter, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=8, fig.align='center'}
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(dNSAF_avg.x) & !is.na(dNSAF_avg.y))
names(data)[4] <- "Value_maize1"
names(data)[5] <- "Value_maize2"
# data <- subset(data, sample %in% c("Silk", "Embryo 20 DAP", "Leaf Zone 2 (Stomatal)"))

subgenome_protein_scatter <-
  ggplot(data, aes(x=Value_maize1, y=Value_maize2)) +
  geom_point(alpha=.1, aes(color = sample)) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  scale_x_log10() + scale_y_log10() +
  labs(
    # title = paste0("Scatterplot of protein abundance for gene pair(s).\nAlpha=.1, reference line"),
    x = "Abundance for Maize1 Gene (log scale)",
    y = "Abundance for Maize2 Gene (log scale)"
  ) +
  theme(legend.position = "none") + facet_wrap(~sample, nrow=6)
  # + annotate("text", size=2, x=10000,y=10, label=paste("pearson: ", round(cor(data$Value_maize1, data$Value_maize2, method = c("pearson")),2)))
plot(subgenome_protein_scatter)
```


# 6 Expression and Abundance are Weakly Correlated

***

```{r subgenome_expression_protein_foldchange_scatter, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=6, fig.height=5, fig.align='center'}
## Start with the paired expression data
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(FPKM_avg.x) & !is.na(FPKM_avg.y))
names(data)[4] <- "FPKM_maize1"
names(data)[5] <- "FPKM_maize2"

## Add paired abundance data
data <-
  data %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize1"="geneID", "sample"="sample")) %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(dNSAF_avg.x) & !is.na(dNSAF_avg.y))
names(data)[6] <- "dNSAF_maize1"
names(data)[7] <- "dNSAF_maize2"

# # Remove pollen
# data <-
#   data %>%
#   subset(!sample %in% c("B73 Mature Pollen"))

data$foldChange_expr <- log2(data$FPKM_maize1) - log2(data$FPKM_maize2)
data$foldChange_abun <- log2(data$dNSAF_maize1) - log2(data$dNSAF_maize2)
subgenome_expression_protein_foldchange_scatter <- ggplot(data, aes(foldChange_expr, foldChange_abun)) +
  geom_point(alpha=.1) +
  xlim(-10,10) + ylim(-10,10) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  # facet_wrap(~sample, nrow=6) +
  labs(
    # title = paste0("Scatterplot of Foldchange in Expression in Maize1 vs.\nFoldchange in Abundance in Maize1\nAlpha=.1, linear fit, n=13413"),
    x = "log2(Expression Maize1) - log2(Expression Maize2)",
    y = "log2(Abundance Maize1) - log2(Abundance Maize2)"
  ) + 
  annotate("text",x=7,y=-8, label=paste("Pearson Cor = ", round(cor(data$foldChange_expr, data$foldChange_abun, method = c("pearson")),2))) + 
  annotate("text",x=7,y=-10, label=paste("Spearman Cor = ", round(cor(data$foldChange_expr, data$foldChange_abun, method = c("spearman")),2)))
plot(subgenome_expression_protein_foldchange_scatter)
```


```{r subgenome_expression_protein_scatter_rep, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=2.5, fig.align='center'}
## Start with the paired expression data
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(FPKM_avg.x) & !is.na(FPKM_avg.y))
names(data)[4] <- "FPKM_maize1"
names(data)[5] <- "FPKM_maize2"

## Add paired abundance data
data <-
  data %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize1"="geneID", "sample"="sample")) %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(dNSAF_avg.x) & !is.na(dNSAF_avg.y))
names(data)[6] <- "dNSAF_maize1"
names(data)[7] <- "dNSAF_maize2"

# Pollen behaves much differently than the rest, lets see
data <- 
  data %>%
  subset(sample %in% c("Primary Root 5 Days", "Mature Leaf 8", "6-7 internode", "B73 Mature Pollen"))

## View as scatter
subgenome_expression_protein_scatter_rep <- ggplot(data, aes(FPKM_maize1, dNSAF_maize1)) +
  geom_point(alpha=.1, aes(color = sample)) +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  # geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  scale_x_log10() + scale_y_log10() +
  theme(legend.position = "none") + facet_wrap(~sample, nrow=1) +
  labs(
    # title = paste0("Scatterplot of Expression in Maize1 vs. Abundance in Maize1\nLog-Log Scale for gene pair(s).\nAlpha=.1, reference line (black), lm fit (blue)"),
    x = "Expression for Maize1 Gene (log scale)",
    y = "Abundance for Maize1 Gene (log scale)"
  ) 
plot(subgenome_expression_protein_scatter_rep)
```

```{r subgenome_expression_protein_scatter, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=8, fig.align='center'}
## Start with the paired expression data
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(FPKM_avg.x) & !is.na(FPKM_avg.y))
names(data)[4] <- "FPKM_maize1"
names(data)[5] <- "FPKM_maize2"

## Add paired abundance data
data <-
  data %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize1"="geneID", "sample"="sample")) %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(dNSAF_avg.x) & !is.na(dNSAF_avg.y))
names(data)[6] <- "dNSAF_maize1"
names(data)[7] <- "dNSAF_maize2"

## View as scatter
subgenome_expression_protein_scatter <- ggplot(data, aes(FPKM_maize1, dNSAF_maize1)) +
  geom_point(alpha=.1, aes(color = sample)) +
  geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95) +
  # geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95) +
  geom_abline(mapping = null, data = null, slope = 1, intercept = 0) +
  scale_x_log10() + scale_y_log10() +
  theme(legend.position = "none") + facet_wrap(~sample, nrow=6) +
  labs(
    # title = paste0("Scatterplot of Expression in Maize1 vs. Abundance in Maize1\nLog-Log Scale for gene pair(s).\nAlpha=.1, reference line (black), lm fit (blue)"),
    x = "Expression for Maize1 Gene (log scale)",
    y = "Abundance for Maize1 Gene (log scale)"
  ) 
plot(subgenome_expression_protein_scatter)
```


# 7 Expression pattern types
functionalDivergence.R

# 8 Misc

***

```{r extra1, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=4, fig.height=4, fig.align='center'}
#Pollen has far fewer cases with exp
## Start with the paired expression data
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(FPKM_avg.x) & !is.na(FPKM_avg.y))
names(data)[4] <- "FPKM_maize1"
names(data)[5] <- "FPKM_maize2"

case_counts <- data %>% group_by(sample) %>% count() %>% arrange(n)
ggplot(case_counts, aes(reorder(sample, n), n)) + geom_col() + theme(axis.text.x=element_text(angle=90,hjust=1))
```

```{r extra2, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=4, fig.height=4, fig.align='center'}
#Pollen has far fewer cases with abun
## Start with the abundance data
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(dNSAF_avg.x) & !is.na(dNSAF_avg.y))
names(data)[4] <- "dNSAF_maize1"
names(data)[5] <- "dNSAF_maize2"

case_counts <- data %>% group_by(sample) %>% count() %>% arrange(n)
ggplot(case_counts, aes(reorder(sample, n), n)) + geom_col() + theme(axis.text.x=element_text(angle=90,hjust=1))
```

```{r extra3, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=4, fig.height=4, fig.align='center'}
#Pollen has far fewer cases with both exp and abun
## Start with the paired expression data
maize.walley.v4mapped.expression.data <- 
  maize.walley.v4mapped.expression %>%
  select(geneID, sample, FPKM_avg)
data <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>%
  select(Maize1, Maize2) %>%
  distinct() %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize1"="geneID")) %>%
  inner_join(maize.walley.v4mapped.expression.data, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(FPKM_avg.x) & !is.na(FPKM_avg.y))
names(data)[4] <- "FPKM_maize1"
names(data)[5] <- "FPKM_maize2"

## Add paired abundance data
data <-
  data %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize1"="geneID", "sample"="sample")) %>%
  inner_join(maize.walley.abundance.v4, by=c("Maize2"="geneID", "sample"="sample")) %>%
  subset(!is.na(dNSAF_avg.x) & !is.na(dNSAF_avg.y))
names(data)[6] <- "dNSAF_maize1"
names(data)[7] <- "dNSAF_maize2"

case_counts <- data %>% group_by(sample) %>% count() %>% arrange(n)
ggplot(case_counts, aes(reorder(sample, n), n)) + geom_col() + theme(axis.text.x=element_text(angle=90,hjust=1))
```

```{r extra4, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=8, fig.align='center'}
# Average expression for genes in each subgenome, biased by sample size.  While other plots show that maize1 dominates maize2 more often in all tissue, maize2 
# for pollen and endosperm crown the average expression is higher in maize2.  Basically, when maize2 wins, it wins by more raw FPKMs.  If we account for all
# genes in the subgenome, not just the pairs, maize2 often has higher average expression.  This is probably due to many very low expressed genes in maize1.
data <- 
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>% #only matching
  # subset(!(Maize1 %in% c("") & Maize2 %in% c(""))) %>% #only nonmatching
  distinct() %>%
  select(Maize1, Maize2) %>%
  left_join(maize.walley.v4mapped.expression, by=c("Maize1"="geneID")) %>%
  rename(FPKM_avg1=FPKM_avg) %>%
  left_join(maize.walley.v4mapped.expression, by=c("Maize2"="geneID", "sample"="sample")) %>%
  rename(FPKM_avg2=FPKM_avg) %>%
  select(Maize1, Maize2, sample, FPKM_avg1, FPKM_avg2) %>%
  # subset(sample %in% c("B73 Mature Pollen")) %>%
  group_by(sample) %>%
  summarize(Mean1=mean(FPKM_avg1, na.rm=TRUE), Mean2=mean(FPKM_avg2, na.rm=TRUE))
data <- 
  data %>%
  gather(subgenome, FPKM, -1)

ggplot(data, aes(x=sample, y=FPKM, fill=subgenome)) + geom_col(position = position_dodge()) + theme(axis.text.x=element_text(angle=90,hjust=1))
```

```{r extra5, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=4, fig.height=4, fig.align='center'}
## In maize1, there are 3672 matching genes and 7927 nonmatching.  Since they tend towards 1 (despite cutoff at 1), more will pull average down.
data <-
  homeologs.pairs %>%
  subset(Maize2 %in% c("")) %>% #only nonmatching
  select(Maize1) %>%
  left_join(maize.walley.v4mapped.expression, by=c("Maize1"="geneID")) %>%
  select(Maize1, sample, FPKM_avg)
temp <-
  homeologs.pairs %>%
  subset(Maize1 != "" & Maize2 != "") %>% #only matching
  select(Maize1) %>%
  left_join(maize.walley.v4mapped.expression, by=c("Maize1"="geneID")) %>%
  select(Maize1, sample, FPKM_avg)

data$status <- "nonMatch"
temp$status <- "match"
data <-
  data %>%
  bind_rows(temp)

ggplot(data, aes(status,log10(FPKM_avg))) +
  geom_boxplot() +
  labs(y = "log10(FPKM)",
       x = "Subgenome",
       title = "Comparison of FPKM Expression between\nGenes in Subgenome 1 and Subgenome 2"
  )
```

```{r extra6, echo=FALSE, results='asis', warning=FALSE, message=FALSE, fig.width=4, fig.height=4, fig.align='center'}
## Not really a plot yet... just sorting types
source("./R/analysis/functionalDivergence_func.R")
"gene"
gene.pair.types <- getTypeSort(homeologs.pairs, maize.walley.v4mapped.expression, 2, .95)
summary(as.factor(gene.pair.types$type))
"protein"
protein.pair.types <- getTypeSortProteins(homeologs.pairs, maize.walley.abundance.v4, 2, .95)
summary(as.factor(protein.pair.types$type))
"agree"
summary(as.factor(intersect(gene.pair.types, protein.pair.types)$type))

data <- 
  data.frame(type=c("gene","protein","agree"),
             values=bind_rows(summary(as.factor(gene.pair.types$type)), 
                              summary(as.factor(protein.pair.types$type)),
                              summary(as.factor(intersect(gene.pair.types, protein.pair.types)$type))
                              )
             ) %>% rename(one_dead=values.1dead, both_dead=values.2dead, ambiguous=values.ambiguous, bed=values.bed, true=values.true)

knitr::kable(data)
# intersect(gene.pair.types[!gene.pair.types$type %in% c("ambiguous"),], protein.pair.types[!protein.pair.types$type %in% c("ambiguous"),])
# #dead expression and still dead abundance (17/297)
# protein.pair.types[protein.pair.types$Maize1 %in% gene.pair.types$Maize1[gene.pair.types$type %in% c("1dead","2dead")],] %>% subset(type %in% c("1dead","2dead"))
```

***

# 9 Wrapup

***

```{r print, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=8, fig.height=8, fig.align='center'}
# # ggsave("Output/Figure 1a.png", plot = chromosome_plot_exp)
# png(filename = "Output/Figure 1a.png", type = "cairo", units = "in", width = 8, height = 8, pointsize = 12, res = 96)
# plot(chromosome_plot_exp)
# dev.off()
# 
# # ggsave("Output/Figure 1b.png", plot = chromosome_plot_abundance)
# png(filename = "Output/Figure 1b.png", type = "cairo", units = "in", width = 8, height = 8, pointsize = 12, res = 96)
# plot(chromosome_plot_abundance)
# dev.off()
# 
# png("Output/Figure 2a.png");
# grid.draw(GOTerms_by_subgenome_a)
# dev.off()
# png("Output/Figure 2b.png");
# grid.draw(GOTerms_by_subgenome_b)
# dev.off()
# 
# # ggsave("Output/Figure 3.png", plot = subgenome_isoforms_scatter)
# png(filename = "Output/Figure 3.png", type = "cairo", units = "in", width = 8, height = 8, pointsize = 12, res = 96)
# plot(subgenome_isoforms_scatter)
# dev.off()
# 
# # ggsave("Output/Figure 4.png", plot = subgenome_expression_scatter_rep)
# png(filename = "Output/Figure 4.png", type = "cairo", units = "in", width = 12, height = 4, pointsize = 12, res = 96)
# plot(subgenome_expression_scatter_rep)
# dev.off()
# 
# # ggsave("Output/Figure 5.png", plot = subgenome_protein_scatter_rep)
# png(filename = "Output/Figure 5.png", type = "cairo", units = "in", width = 12, height = 4, pointsize = 12, res = 96)
# plot(subgenome_protein_scatter_rep)
# dev.off()
# 
# # ggsave("Output/Figure 6.png", plot = subgenome_expression_protein_scatter_rep)
# png(filename = "Output/Figure 6.png", type = "cairo", units = "in", width = 12, height = 4, pointsize = 12, res = 96)
# plot(subgenome_expression_protein_scatter_rep)
# dev.off()
# 
# ggsave("Output/Figure S1.png", plot = subgenome_expression_scatter)
# ggsave("Output/Figure S2.png", plot = subgenome_protein_scatter)
# ggsave("Output/Figure S3.png", plot = subgenome_expression_protein_scatter)
# ggsave("Output/Figure S4.png", plot = subgenome_expression_protein_foldchange_scatter)
# 
# list.of.subgenome.assignments <- subgenome %>% select(gene2, subgenome) %>% rename(B73.v4.ID=gene2, Subgenome=subgenome)
# write.csv(list.of.subgenome.assignments, "list.of.subgenome.assignments.csv")
# 
# write.csv(homeologs.pairs %>% subset(Maize1 != "" & Maize2 != "") %>% select(Maize1, Maize2) %>% rename(B73.v4.ID.Maize1=Maize1, B73.v4.ID.Maize2=Maize2), "list.of.retained.duplicate.pairs")
```

Report generated on: `r as.character(format(Sys.Date(), format="%B %d, %Y"))`
